### PR TITLE
Fix promotion rules for GenericNonlinearExpr

### PR DESCRIPTION
**Pull request by odow**
Closes https://github.com/jump-dev/JuMP.jl/issues/3482

https://github.com/jump-dev/JuMP.jl/actions/runs/6067289679

### CODE REVIEW

**Comment by odow** [comment]
This fixes the UnitJuMP test failure, although UnitJuMP still commits type piracy.

**Review by mlubin — APPROVED** [review]

**Comment by odow** [comment]
It's still possible to generate `Vector{AbstractJuMPScalar}`:
```Julia
julia> using JuMP

julia> model = Model();

julia> @variable(model, x);

julia> y = [i == 1 ? x : sin(x) for i in 1:2]
2-element Vector{AbstractJuMPScalar}:
 x
 sin(x)

julia> @objective(model, Min, [i == 1 ? x : sin(x) for i in 1:2])
ERROR: MethodError: no method matching moi_function(::Vector{AbstractJuMPScalar})
Closest candidates are:
  moi_function(::AbstractConstraint) at /Users/oscar/.julia/dev/JuMP/src/constraints.jl:550
  moi_function(::AbstractVariableRef) at /Users/oscar/.julia/dev/JuMP/src/variables.jl:575
  moi_function(::Vector{var"#s65"} where var"#s65"<:AbstractVariableRef) at /Users/oscar/.julia/dev/JuMP/src/variables.jl:584
  ...
Stacktrace:
 [1] set_objective_function(model::Model, func::Vector{AbstractJuMPScalar})
   @ JuMP ~/.julia/dev/JuMP/src/objective.jl:146
 [2] set_objective(model::Model, sense::MathOptInterface.OptimizationSense, func::Vector{AbstractJuMPScalar})
   @ JuMP ~/.julia/dev/JuMP/src/objective.jl:175
 [3] macro expansion
   @ ~/.julia/dev/JuMP/src/macros.jl:1850 [inlined]
 [4] top-level scope
   @ REPL[5]:1

julia> @objective(model, Min, y)
ERROR: MethodError: no method matching moi_function(::Vector{AbstractJuMPScalar})
```

So perhaps we need a better error for the fallback.

**Comment by odow** [comment]
Thoughts @pulsipher?

**Comment by odow** [comment]
@trulsf fixed UnitJuMP, but I think this is the right thing regardless. We can always add back support for `Vector{AbstractJuMPScalar}` later.

File: src/nlp_expr.jl [hunk 1]
```diff
@@ -1097,18 +1111,15 @@ function jump_function(
...
     return MOI.VectorNonlinearFunction
 end
 
-function moi_function(f::Vector{<:AbstractJuMPScalar})
+function moi_function(f::AbstractVector{<:GenericNonlinearExpr})
     return MOI.VectorNonlinearFunction(f)
 end
 
-function MOI.VectorNonlinearFunction(f::Vector{<:AbstractJuMPScalar})
+function MOI.VectorNonlinearFunction(f::AbstractVector{<:GenericNonlinearExpr})
```
Comment (blegat) [inline src/nlp_expr.jl:1122]: I don't agree. Here, you are explicit about the return type so it should work with a vector of anything

Comment (blegat) [reply src/nlp_expr.jl:1122]: But it can be added later in a non breaking way so I won't dispute this

**Review by blegat — APPROVED** [review]

File: src/nlp_expr.jl (hunk 1 above)
Comment (odow) [reply src/nlp_expr.jl:1122]: Ah true. This is reasonable.

Comment (odow) [inline src/nlp_expr.jl:1122]:
```suggestion
function MOI.VectorNonlinearFunction(f::Vector{<:AbstractJuMPScalar})
```

**Pull request merged**